src/main/java/io/confluent/kafkarest/SimpleConsumerManager.java

```diff
@@ -0,0 +1,228 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.confluent.kafkarest;
+
+import io.confluent.kafka.serializers.KafkaAvroDecoder;
+import io.confluent.kafkarest.converters.AvroConverter;
+import io.confluent.kafkarest.entities.AvroConsumerRecord;
+import io.confluent.kafkarest.entities.BinaryConsumerRecord;
+import io.confluent.kafkarest.entities.ConsumerRecord;
+import io.confluent.kafkarest.entities.EmbeddedFormat;
+import io.confluent.rest.exceptions.RestException;
+import io.confluent.rest.exceptions.RestServerErrorException;
+import kafka.api.PartitionFetchInfo;
+import kafka.cluster.Broker;
+import kafka.common.TopicAndPartition;
+import kafka.javaapi.FetchRequest;
+import kafka.javaapi.FetchResponse;
+import kafka.javaapi.message.ByteBufferMessageSet;
+import kafka.message.MessageAndMetadata;
+import kafka.message.MessageAndOffset;
+import kafka.serializer.Decoder;
+import kafka.serializer.DefaultDecoder;
+import kafka.utils.VerifiableProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.core.Response;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class SimpleConsumerManager {
+
+  private static final Logger log = LoggerFactory.getLogger(SimpleConsumerManager.class);
+
+  private final int maxPoolSize;
+  private final MetadataObserver mdObserver;
+  private final SimpleConsumerFactory simpleConsumerFactory;
+
+  private final ConcurrentMap<Broker, SimpleConsumerPool> simpleConsumersPools;
+
+  private AtomicInteger correlationId = new AtomicInteger(0);
+
+  private final Decoder<Object> avroDecoder;
+  private final Decoder<byte[]> binaryDecoder;
+
+  public SimpleConsumerManager(final KafkaRestConfig config,
+                               final MetadataObserver mdObserver,
+                               final SimpleConsumerFactory simpleConsumerFactory) {
+
+    this.mdObserver = mdObserver;
+    this.simpleConsumerFactory = simpleConsumerFactory;
+
+    maxPoolSize = config.getInt(KafkaRestConfig.SIMPLE_CONSUMER_POOL_SIZE_CONFIG);
+    simpleConsumersPools = new ConcurrentHashMap<Broker, SimpleConsumerPool>();
+
+    // Load decoders
+    Properties props = new Properties();
+    props.setProperty("schema.registry.url",
+        config.getString(KafkaRestConfig.SCHEMA_REGISTRY_URL_CONFIG));
+    avroDecoder = new KafkaAvroDecoder(new VerifiableProperties(props));
+
+    binaryDecoder = new DefaultDecoder(new VerifiableProperties());
+  }
+
+  private SimpleConsumerPool createSimpleConsumerPool() {
+    return new SimpleConsumerPool(maxPoolSize, simpleConsumerFactory);
+  }
+
+  private SimpleFetcher getSimpleFetcher(final Broker broker) {
+    if (!simpleConsumersPools.containsKey(broker)) {
+      simpleConsumersPools.put(broker, createSimpleConsumerPool());
```

Comment:
There's a race here since two threads could get past the conditional and then both put(). The subsequent get() isn't enough to make sure it's correct because both threads could get past the containsKey() check, then thread A could put() and get(), then thread B could do the same.

This is a pretty common error with ConcurrentHashMaps. You can either use the same approach but use putIfAbsent instead of put (at the cost of allocating, but possibly not using a SimpleConsumerPool), or use a regular HashMap and handle synchronization manually.

In this case, the extra pool and anything it allocated should just be garbage collected after the request that creates it completes, but it'd be better to just fix this to be completely correct.